core/src/test/resources/org/apache/calcite/test/SqlToRelConverterTest.xml

```diff
@@ -867,6 +867,17 @@ LogicalProject(ZIP=[$4])
   LogicalFilter(condition=[=($3, 'abc':VARCHAR(20))])
     LogicalProject(EMPNO=[$0], STREET=[$1.STREET], CITY=[$1.CITY], ZIP=[$1.ZIP], STATE=[$1.STATE], STREET5=[$2.STREET], CITY6=[$2.CITY], ZIP7=[$2.ZIP], STATE8=[$2.STATE])
       LogicalTableScan(table=[[CATALOG, SALES, EMP_ADDRESS]])
+]]>
+        </Resource>
+    </TestCase>
+    <TestCase name="testSelectNestedColumnType">
+        <Resource name="plan">
+            <![CDATA[
+LogicalProject(COORD=[$0])
```

Comment:
Hey @my7ym , I went quickly over the PR but I was a bit confused with this projection. Based on the query the final projection should be a structured type (coord) and here it appears that the final projection is just an int type (coord.x). Do you have an explanation for this?

Reply:
Appreciate your detailed review @zabetak and I was confused too. 

It was because the test utility converts the SqlNode to RelNode on this line
https://github.com/apache/calcite/blob/ebafff1ede5921901b322872693bf1dbdf09fc8d/core/src/test/java/org/apache/calcite/test/SqlToRelTestBase.java#L723

And routes to 
https://github.com/apache/calcite/blob/3fa29455664bec0056c436491b369e0cd72242ea/core/src/main/java/org/apache/calcite/rel/RelRoot.java#L149

In this function, the original top level project
```
LogicalProject(COORD=[$4], COORD1=[$5], COORD2=[$6])
```
is identified as not trivial and RelRoot creates an extra layer of Project on top of it.

This behavior should be in codebase for a long time and orthogonal with my fix. And I am wondering what's the original intention to PROJECT a root. (When I am using RelRoot, I just RelRoot.rel to get the relation).

Reply:
The root is used to remove redundant fields that may be introduced for various reasons. For instance, the fields participating in an ORDER BY clause will appear in the project even if they don't appear in the SELECT clause; in that case the root will be used to remove these fields before returning the answer to the user. 

As you say the problem is probably in another part of the code, but I would prefer if we could find another test that reproduces the problem and also has a correct plan. Apparently, thats not easy since I tried but with no success till now. I will give it some more thought and if we don't find a better test, I will merge this as is ;)

In general, this use-case makes me wonder a bit. What is the purpose of the type flattener if at the end we need to return an unflattened type.

Reply:
@zabetak Got your point.

The root actually got expanded like:
```
  LogicalProject(COORD=[$4], COORD1=[$5], COORD2=[$6])
    LogicalFilter(condition=[=($4, 1)])
      LogicalProject(CONTACTNO=[$0], FNAME=[$1], LNAME=[$2], EMAIL=[$3], X=[$4.X], Y=[$4.Y], unit=[$4.unit], M=[$5.M], A=[$5.SUB.A], B=[$5.SUB.B])
        LogicalTableScan(table=[[CATALOG, CUSTOMER, CONTACT_PEEK]])
```
But the RelRoot.project messed it up :(

And I like the original test case better because it's the minimum one which could test this failure.

But I think your point also makes sense. So I managed to make up a test to cover this issue. (And this test case is extremely similar to my project)

Let me know what you think man.

Reply:
Thanks man! Let me know when you approve. I will stash commits and append my name in the commit message as last time.